File: gitlab-nic/code-samples-main/API/Go/afnic.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
)

func headers() map[string]string {
	filename := os.Getenv("HOME") + "/.afnic-api"
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	rd := bufio.NewReader(file)
	login, err := rd.ReadString('\n')
	if err != nil {
		panic(err)
	}
	login = login[:len(login)-1]
	password, err := rd.ReadString('\n')
	if err != nil {
		panic(err)
	}
	password = password[:len(password)-1]
	values := url.Values{"client_id": {"registrars-api-client"},
		"username": {login}, "password": {password},
		"grant_type": {"password"}}
	response, err := http.PostForm("https://login-sandbox.nic.fr/auth/realms/fr/protocol/openid-connect/token",
		values)
	if err != nil {
		panic(err)
	}
	defer response.Body.Close()
	if response.StatusCode != 200 {
		panic(fmt.Sprintf("Wrong HTTP response for token: %s", response.Status))
	}
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		panic(err)
	}
	type Result struct {
		Access_token string /* We ignore the other members */
	}
	var result Result
	err = json.Unmarshal(body, &result)
	if err != nil {
		panic(err)
	}
	return map[string]string{"Content-Type": "application/json",
		"Accept": "application/json",
		"Extensions": "FRNIC_V2, SECDNS_V1_1",
		"Authorization": fmt.Sprintf("Bearer %s", result.Access_token)}
}

File: gitlab-nic/code-samples-main/API/Elixir/lib/afnic.ex
defmodule Afnic do
  @filename System.get_env("HOME") <> "/.afnic-api"

  def get_token() do
    {login, password} =
      case File.open(@filename, [:read, :utf8]) do
        {:ok, handle} ->
          {String.trim(IO.read(handle, :line)), String.trim(IO.read(handle, :line))}

        {:error, reason} ->
          raise "Boom #{@filename}: #{reason}"
      end

    url = "https://login-sandbox.nic.fr/auth/realms/fr/protocol/openid-connect/token"

    body =
      {:form,
       [
         client_id: "registrars-api-client",
         username: login,
         password: password,
         grant_type: "password"
       ]}

    headers = [{"Accept", "application/json"}]
    result = HTTPoison.post!(url, body, headers, [])

    case result do
      %HTTPoison.Response{status_code: 200, body: body} ->
        response = Jason.decode!(body)
        response["access_token"]

      other ->
        raise "Boom (#{inspect(other)})"
    end
  end

  def get_headers() do
    token = get_token()

    [
      {"Accept", "application/json"},
      {"Content-Type", "application/json"},
      {"Extensions", "FRNIC_V2, SECDNS_V1_1"},
      {"Authorization", "Bearer #{token}"}
    ]
  end
end
